### PR TITLE
dnn: add shared fastNorm kernel for mvn, instance norm and layer norm

### PR DESCRIPTION
**Pull request by fengyuentau**
Relates https://github.com/opencv/opencv/pull/24378#issuecomment-1756906570

TODO:

- [x] add fastNorm
- [x] refactor layer norm with fastNorm
- [x] refactor mvn with fastNorm
- [ ] add onnx mvn in importer (in a new PR?)
- [ ] refactor instance norm with fastNorm (in another PR https://github.com/opencv/opencv/pull/24378, need to merge this one first though)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by fengyuentau** [comment]
It seems there are a few caffe models using mvn layer. I cannot find one in opencv_extra.

**Comment by dkurt** [comment]
@fengyuentau, [AlexNet](https://github.com/opencv/opencv_extra/blob/522b48c1649bc6893a72a6298c76ddbd838ac6dd/testdata/dnn/download_models.py#L226) maybe?

**Comment by fengyuentau** [comment]
> @fengyuentau, [AlexNet](https://github.com/opencv/opencv_extra/blob/522b48c1649bc6893a72a6298c76ddbd838ac6dd/testdata/dnn/download_models.py#L226) maybe?

Seems like a no, see below:

image

**Comment by dkurt** [comment]
@fengyuentau, sorry, it's LRN.

There are tests which use MVN:
```
[  FAILED  ] Test_Caffe_layers.MVN/0, where GetParam() = OCV/CPU
[  FAILED  ] Test_Caffe_layers.BatchNorm/0, where GetParam() = OCV/CPU
[  FAILED  ] Test_ONNX_layers.InstanceNorm/0, where GetParam() = OCV/CPU
[  FAILED  ] Test_TensorFlow_layers.batch_norm_10/0, where GetParam() = OCV/CPU
[  FAILED  ] Test_TensorFlow_layers.batch_norm_11/0, where GetParam() = OCV/CPU
[  FAILED  ] Test_TensorFlow_layers.batch_norm_13/0, where GetParam() = OCV/CPU
[  FAILED  ] Test_Torch_layers.run_batch_norm/0, where GetParam() = OCV/CPU
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = OCV/CPU
```

**Comment by dkurt** [comment]
@fengyuentau, exactly. This is for channel-wise fusion.

**Comment by fengyuentau** [comment]
> @fengyuentau, exactly. This is for channel-wise scale/bias.

Thank you for the clarification. This PR is all set for review. Most of the tests are green on my mac m1 except for something wrong with `Test_Torch_nets.FastNeuralStyle_accuracy`.

**Comment by fengyuentau** [comment]
Should we just tune the threshold?

[ubuntu arm64:](https://github.com/opencv/opencv/actions/runs/6545660238/job/17774610114?pr=24409)
```
[ RUN      ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = OCV/CPU
/home/ci/opencv/modules/dnn/test/test_common.impl.hpp:79: Failure
Expected: (normInf) <= (lInf), actual: 1.14948 vs 1.11
  |ref| = 255
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = OCV/CPU (279 ms)
```

[ubuntu x64:](https://github.com/opencv/opencv/actions/runs/6545660238/job/17774611445?pr=24409)
```
[ RUN      ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = OCV/CPU
/home/ci/opencv/modules/dnn/test/test_common.impl.hpp:79: Failure
Expected: (normInf) <= (lInf), actual: 1.15276 vs 1.11
  |ref| = 255
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = OCV/CPU (335 ms)
```

[windows10 x64 vulkan:](https://github.com/opencv/opencv/actions/runs/6545660238/job/17774614462?pr=24409)
```
[ RUN      ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = VKCOM/VULKAN
C:\GHA-OCV-2\_work\opencv\opencv\opencv\modules\dnn\test\test_common.impl.hpp(79): error: Expected: (normInf) <= (lInf), actual: 1.1487 vs 1.11
  |ref| = 255
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = VKCOM/VULKAN (686 ms)
[ RUN      ] Test_Torch_nets.FastNeuralStyle_accuracy/1, where GetParam() = OCV/CPU
C:\GHA-OCV-2\_work\opencv\opencv\opencv\modules\dnn\test\test_common.impl.hpp(79): error: Expected: (normInf) <= (lInf), actual: 1.15276 vs 1.11
  |ref| = 255
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/1, where GetParam() = OCV/CPU (572 ms)
```

[macos arm64:](https://github.com/opencv/opencv/actions/runs/6545660238/job/17774607389?pr=24409)
```
[ RUN      ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = OCV/CPU
/Users/opencv-cn/GHA-OCV-1/_work/opencv/opencv/opencv/modules/dnn/test/test_common.impl.hpp:79: Failure
Expected: (normInf) <= (lInf), actual: 1.15292 vs 1.11
  |ref| = 255
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/0, where GetParam() = OCV/CPU (350 ms)
[ RUN      ] Test_Torch_nets.FastNeuralStyle_accuracy/1, where GetParam() = OCV/CPU_FP16
/Users/opencv-cn/GHA-OCV-1/_work/opencv/opencv/opencv/modules/dnn/test/test_common.impl.hpp:76: Failure
Expected: (normL1) <= (l1), actual: 0.633552 vs 0.62
  |ref| = 255
[  FAILED  ] Test_Torch_nets.FastNeuralStyle_accuracy/1, where GetParam() = OCV/CPU_FP16 (336 ms)
```

File: modules/dnn/src/layers/cpu_kernels/fast_norm.cpp [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+            float mean = 0.f, mean_square = 0.f;
+            for (int j = 0; j < norm_size; j++) {
+                float v = x[j];
+                mean += v;
+                mean_square += v * v;
+            }
+
+            mean *= inv_norm_size;
+            mean_square = std::sqrt(std::max(0.f, mean_square * inv_norm_size - mean * mean) + epsilon);
+            float inv_stdev = normalize_variance ? 1.f / mean_square : 1.f;
```
Comment (dkurt) [inline modules/dnn/src/layers/cpu_kernels/fast_norm.cpp:36]: Seems like we don't need to compute `mean_square` if `normalize_variance=false`

File: modules/dnn/src/layers/layer_norm.cpp [hunk 2]
```diff
@@ -160,10 +79,16 @@ class LayerNormLayerImpl CV_FINAL : public LayerNormLayer
         inputs_arr.getMatVector(inputs);
         outputs_arr.getMatVector(outputs);
 
-        if (hasBias) {
-            LayerNormInvoker<true>::run(inputs[0], inputs[1], &inputs[2], outputs[0], axis, epsilon);
+        const auto &input = inputs[0];
+        const auto &scale = inputs[1];
+        const auto scale_shape = shape(scale);
```
Comment (dkurt) [inline modules/dnn/src/layers/layer_norm.cpp:84]: unused

File: modules/dnn/src/layers/cpu_kernels/fast_norm.cpp [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+            float mean = 0.f, mean_square = 0.f;
+            for (int j = 0; j < norm_size; j++) {
+                float v = x[j];
+                mean += v;
+                mean_square += v * v;
+            }
+
+            mean *= inv_norm_size;
+            mean_square = std::sqrt(std::max(0.f, mean_square * inv_norm_size - mean * mean) + epsilon);
```
Comment (dkurt) [inline modules/dnn/src/layers/cpu_kernels/fast_norm.cpp:35]: Why `std::max` is called?

File: modules/dnn/src/layers/cpu_kernels/fast_norm.cpp (hunk 1 above)
Comment (fengyuentau) [reply modules/dnn/src/layers/cpu_kernels/fast_norm.cpp:36]: It is true but given the fact that `normalize_variance=false` is rare case so I guess this implementation is acceptable.

File: modules/dnn/src/layers/cpu_kernels/fast_norm.cpp (hunk 3 above)
Comment (fengyuentau) [reply modules/dnn/src/layers/cpu_kernels/fast_norm.cpp:35]: See https://github.com/opencv/opencv/pull/23047#discussion_r1082139598

File: modules/dnn/src/layers/layer_norm.cpp (hunk 2 above)
Comment (fengyuentau) [reply modules/dnn/src/layers/layer_norm.cpp]: Done.

File: modules/dnn/include/opencv2/dnn/all_layers.hpp [hunk 4]
```diff
@@ -1127,7 +1127,6 @@ CV__DNN_INLINE_NS_BEGIN
     class CV_EXPORTS LayerNormLayer : public Layer
     {
     public:
-        bool hasBias;
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/all_layers.hpp:1130]: Breaks ABI check

Comment (fengyuentau) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp:1130]: Yes, I know this. Actually I want to drop this variable. If we preserve ABI compatibility, we can keep it but not using it.

**Review by dkurt — APPROVED** [review]

File: modules/dnn/include/opencv2/dnn/all_layers.hpp (hunk 4 above)
Comment (fengyuentau) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp:1130]: Reverted this change. Lets keep it for compatiblity.

File: modules/dnn/include/opencv2/dnn/all_layers.hpp [hunk 5]
```diff
@@ -1127,6 +1127,7 @@ CV__DNN_INLINE_NS_BEGIN
     class CV_EXPORTS LayerNormLayer : public Layer
     {
     public:
+        bool hasBias; // Deprecated, preserve for compatibility
```
Comment (dkurt) [inline modules/dnn/include/opencv2/dnn/all_layers.hpp:1130]: `CV_DEPRECATED_EXTERNAL bool hasBias`

Comment (fengyuentau) [reply modules/dnn/include/opencv2/dnn/all_layers.hpp]: Thank you!

**Pull request merged**